### PR TITLE
utils + azure: vscode.dev migration work

### PR DESCRIPTION
**Pull request by nturinski**
TODO: There's still some commented out code that I need to fix before we can merge this, just getting the PR up

### CODE REVIEW

File: utils/src/utils/openUrl.ts [hunk 1]
```diff
@@ -3,11 +3,9 @@
...
 
-import open = require("open");
+import * as vscode from 'vscode';
 
 export async function openUrl(url: string): Promise<void> {
     // Using this functionality is blocked by https://github.com/Microsoft/vscode/issues/85930
-    // await vscode.commands.executeCommand('vscode.open', vscode.Uri.parse(url));
-
-    await open(url);
+    await vscode.commands.executeCommand('vscode.open', vscode.Uri.parse(url));
```
Comment (bwateratmsft) [inline utils/src/utils/openUrl.ts:10]: Might need to make changes in the default webpack config in the dev package too, with `open` removed

File: azure/src/index.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@
  *--------------------------------------------------------------------------------------------*/
 
 export * from './createAzureClient';
+export { registerAzureUtilsExtensionVariables } from './extensionVariables';
```
Comment (alexweininger) [inline azure/src/index.ts:7]: nit: Undo this change since it's just reordering the imports?

File: azure/src/azure-account.api.d.ts [hunk 3]
```diff
@@ -4,61 +4,61 @@
  *--------------------------------------------------------------------------------------------*/
```
Comment (alexweininger) [inline azure/src/azure-account.api.d.ts:4]: nit: revert the whitespace changes in this file.

File: azure/package.json [hunk 4]
```diff
@@ -35,19 +35,23 @@
...
+        "@azure/arm-storage": "^18.0.0",
         "@azure/arm-storage-profile-2020-09-01-hybrid": "^2.0.0",
-        "@azure/ms-rest-js": "^2.2.1",
-        "@microsoft/vscode-azext-utils": "^0.4.3",
+        "@azure/core-client": "^1.6.0",
+        "@azure/core-rest-pipeline": "^1.9.0",
+        "@microsoft/vscode-azext-utils": "0.4.3-alpha",
         "semver": "^7.3.7",
+        "uuid": "^9.0.0",
+        "uuidv4": "^6.2.13",
```
Comment (alexweininger) [inline azure/package.json:45]: Why do we need both of these?

File: azure/package.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "@microsoft/vscode-azext-azureutils",
     "author": "Microsoft Corporation",
-    "version": "0.3.8",
+    "version": "0.3.9",
```
Comment (alexweininger) [inline azure/package.json:4]: These changes may warrant a minor version bump. Specifically around the client/pipeline stuff.

Part of me wants to release a 1.0.0 of the tools package so we can actually use major/minor/patch instead of just minor/patch.

File: utils/package.json [hunk 6]
```diff
@@ -34,18 +34,20 @@
...
         "@microsoft/vscode-azureresources-api": "^2.0.2",
         "@vscode/extension-telemetry": "^0.6.2",
+        "crypto-randomuuid": "^1.0.0",
         "dayjs": "^1.11.2",
         "escape-string-regexp": "^2.0.0",
-        "html-to-text": "^8.2.0",
-        "open": "^8.0.4",
         "semver": "^7.3.7",
+        "text-encoding": "^0.7.0",
+        "uuid": "^9.0.0",
```
Comment (alexweininger) [inline utils/package.json:42]: We need both `crypto-randomuuid` and `uuid`?

File: utils/src/createExperimentationService.ts [hunk 7]
```diff
@@ -1,135 +1,135 @@
...
-
-    /**
-     * Implements a telemetry handler that adds the shared properties to the event
-     * @param actionContext The action context
-     */
-    public handleTelemetry(actionContext: IActionContext): void {
-        Object.assign(actionContext.telemetry.properties, this.sharedProperties);
-    }
-}
+// /*---------------------------------------------------------------------------------------------
```
Comment (alexweininger) [inline utils/src/createExperimentationService.ts:1]: What's the plan for this?

File: utils/src/createTelemetryReporter.ts [hunk 8]
```diff
@@ -3,9 +3,9 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
+import TelemetryReporter from '@vscode/extension-telemetry';
```
Comment (alexweininger) [inline utils/src/createTelemetryReporter.ts:6]: nit: undo reordering

File: utils/src/registerReportIssueCommand.ts [hunk 9]
```diff
@@ -3,16 +3,16 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import * as dayjs from 'dayjs';
-// eslint-disable-next-line import/no-internal-modules
-import * as relativeTime from 'dayjs/plugin/relativeTime';
+// import * as dayjs from 'dayjs';
```
Comment (alexweininger) [inline utils/src/registerReportIssueCommand.ts:6]: We can't get `dayjs` to work? 😢

File: utils/src/utils/openUrl.ts (hunk 1 above)
Comment (alexweininger) [inline utils/src/utils/openUrl.ts:10]: I'm sorry I think this has been discussed already, but why can't we use `vscode.env.openExternal` here?

Comment (bwateratmsft) [reply utils/src/utils/openUrl.ts:10]: VSCode is over-encoding some parts of a URI: https://github.com/Microsoft/vscode/issues/85930

However, the workaround suggested--pass a `string` and use `as unknown as vscode.Uri` to placate Typescript--should work. The code definitely very deliberately supports strings being passed in, even if the type declaration doesn't say so.

I think we should change this `openUrl` utility method to accept either `string` or `vscode.Uri`, and pass unchanged.

Comment (nturinski) [reply utils/src/utils/openUrl.ts:10]: I made this change in resources, but I think that I'll just do it here and delete it everywhere else?

File: azure/src/index.ts (hunk 2 above)
Comment (nturinski) [reply azure/src/index.ts:7]: I mean, it was in the wrong spot before anyway. It's in alphabetical order now.

File: azure/src/azure-account.api.d.ts (hunk 3 above)
Comment (nturinski) [reply azure/src/azure-account.api.d.ts:4]: I think the spaces were wrong originally. Everything is indented twice for some reason.

File: azure/package.json (hunk 5 above)
Comment (nturinski) [reply azure/package.json:4]: I wouldn't mind. We don't really have a "GA" process for these packages so we could just go ahead and do it.

File: azure/src/index.ts (hunk 2 above)
Comment (alexweininger) [reply azure/src/index.ts:7]: Ok fine. I'm just being nit-picky about this PR touching all these extra files. Kinda messes with the git blame. I'd prefer if we fixed all the imports in a separate PR but it's fine as is.

File: utils/src/createExperimentationService.ts (hunk 7 above)
Comment (nturinski) [reply utils/src/createExperimentationService.ts:1]: Looks like I was able to just uncomment it and it's working. I can't remember what was broken before.

Comment (alexweininger) [reply utils/src/createExperimentationService.ts:1]: 😄

File: utils/src/registerReportIssueCommand.ts (hunk 9 above)
Comment (nturinski) [reply utils/src/registerReportIssueCommand.ts:6]: It didn't have a browser entry point in their `package.json` so I may have just assumed. I'll see if it works-- the browser entry point may only specifically be a VS Code for the Web thing,

File: utils/src/createTelemetryReporter.ts (hunk 8 above)
Comment (alexweininger) [reply utils/src/createTelemetryReporter.ts:6]: Ok, then just ignore my nits about undoing the import reordering.

File: azure/src/azure-account.api.d.ts (hunk 3 above)
Comment (bwateratmsft) [reply azure/src/azure-account.api.d.ts:4]: It was tabs before, IIRC, and every time we get near it, it's trying to change it to spaces. Not even worth fighting with it...

File: azure/test/request.test.ts [hunk 10]
```diff
@@ -115,38 +115,22 @@ suite('request', () => {
...
-
-        const request = new WebResource(url);
-        request.operationSpec = { httpMethod: "GET", responses: { 200: {}, default: {} }, serializer: new Serializer() };
-        const client = await createGenericClient(await createTestActionContext(), undefined);
-        await assertThrowsAsync(async () => await client.sendRequest(request), /oops/);
-    });
-
-    test('operationSpec with expected error', async () => {
-        testResponses = [{ status: 404, body: 'oops' }];
+    //test('operationSpec with unexpected  error', async () => {
```
Comment (nturinski) [inline azure/test/request.test.ts:118]: I'm not sure if we're still using operationSpec errors because it was our StatusCodePolicy pipeline controlling that. There are still some pipelines that I need to investigate into adding back in.

File: utils/src/registerReportIssueCommand.ts (hunk 9 above)
Comment (alexweininger) [reply utils/src/registerReportIssueCommand.ts]: `browser` is a standardized property, however it's not required and I think a lot of packages haven't switched over to it yet.

File: dev/src/webpack/getDefaultWebpackConfig.ts [hunk 11]
```diff
@@ -211,6 +209,10 @@ export function getDefaultWebpackConfig(options: DefaultWebpackOptions): webpack
                     "async_hooks": false,
                     "child_process": false,
                     "fs": false,
+                    'html-to-text': false,
```
Comment (nturinski) [inline dev/src/webpack/getDefaultWebpackConfig.ts:212]: This package is used to parse errors. I'm not sure how to really test it (because I don't see html errors often) but it's wrapped in a try/catch so I don't think that this should cause any exceptions to be thrown in a vscode.dev environment.

Comment (bwateratmsft) [reply dev/src/webpack/getDefaultWebpackConfig.ts:212]: Hasn't this one given us CG alerts before? Just to parse errors feels like a pretty flimsy reason to keep it...

Comment (alexweininger) [reply dev/src/webpack/getDefaultWebpackConfig.ts:212]: It's only used to parse error messages in case the error message is an html document, which seems somewhat common for Azure errors. See https://github.com/microsoft/vscode-azuretools/pull/360

Comment (alexweininger) [reply dev/src/webpack/getDefaultWebpackConfig.ts:212]: Also, I couldn't quickly find any alternatives that are browser compatible, so 🤷

Comment (bwateratmsft) [reply dev/src/webpack/getDefaultWebpackConfig.ts:212]: Oh the irony...can't find a browser-compatible package to parse HTML

File: utils/src/registerReportIssueCommand.ts (hunk 9 above)
Comment (nturinski) [reply utils/src/registerReportIssueCommand.ts]: Yeah, good news, I'm pretty sure that it does work.

File: azure/package.json [hunk 12]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "@microsoft/vscode-azext-azureutils",
     "author": "Microsoft Corporation",
-    "version": "0.3.8",
+    "version": "1.0.0",
```
[the remark this answers is not in the document]
Comment (nturinski) [reply azure/package.json:4]: lol I thought we agreed to this?!?!

File: azure/test/request.test.ts [hunk 13]
```diff
@@ -115,38 +115,22 @@ suite('request', () => {
...
 
-    test('operationSpec with unexpected  error', async () => {
-        testResponses = [{ status: 404, body: 'oops' }];
-
-        const request = new WebResource(url);
-        request.operationSpec = { httpMethod: "GET", responses: { 200: {}, default: {} }, serializer: new Serializer() };
-        const client = await createGenericClient(await createTestActionContext(), undefined);
-        await assertThrowsAsync(async () => await client.sendRequest(request), /oops/);
-    });
+    // test('operationSpec with unexpected  error', async () => {
```
Comment (alexweininger) [inline azure/test/request.test.ts:118]: Can we uncomment these tests?

File: azure/package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexweininger) [reply azure/package.json:4]: The gif was supposed to relay celebration, but looks like I picked the wrong gif. I really wanted a Kramer gif. I was lazy and didn't take the time to pick the right gif. That's on me.

File: azure/index.d.ts [hunk 14]
```diff
@@ -377,18 +362,19 @@ export type AzExtGenericClientInfo = AzExtGenericCredentials | { credentials: Az
  * 2. Uses resourceManagerEndpointUrl to support sovereigns (if clientInfo corresponds to an Azure environment)
  * @param clientInfo The client/credentials info or `undefined` if no credentials are needed
  */
-export function createGenericClient(context: IActionContext, clientInfo: AzExtGenericClientInfo): Promise<ServiceClient>;
+export declare function createGenericClient(context: IActionContext, clientInfo: AzExtGenericClientInfo | undefined): Promise<ServiceClient>;
```
Comment (alexweininger) [inline azure/index.d.ts:365]:
```suggestion
export declare function createGenericClient(context: IActionContext, clientInfo?: AzExtGenericClientInfo): Promise<ServiceClient>;
```

File: azure/src/createAzureClient.ts [hunk 15]
```diff
@@ -35,131 +33,121 @@ export function parseClientContext(clientContext: InternalAzExtClientContext): I
...
+    const client = await createGenericClient(context, clientInfo, { noRetryPolicy: true });
     return await client.sendRequest(request);
 }
 
 interface IGenericClientOptions {
     noRetryPolicy?: boolean;
 }
 
-export async function createGenericClient(context: IActionContext, clientInfo: types.AzExtGenericClientInfo, options?: IGenericClientOptions): Promise<ServiceClient> {
+export async function createGenericClient(context: IActionContext, clientInfo: types.AzExtGenericClientInfo | undefined, options?: IGenericClientOptions): Promise<ServiceClient> {
```
Comment (alexweininger) [inline azure/src/createAzureClient.ts:74]:
```suggestion
export async function createGenericClient(context: IActionContext, clientInfo?: types.AzExtGenericClientInfo, options?: IGenericClientOptions): Promise<ServiceClient> {
```

File: azure/test/request.test.ts (hunk 13 above)
Comment (nturinski) [reply azure/test/request.test.ts:118]: Not yet-- I need to either edit them or manipulate with the pipelines a bit more to see why they are still failing. 

But the way that the error is shaped in the tests don't really make sense to me and I don't see any pipeline that would fix this.

File: azure/index.d.ts (hunk 14 above)
Comment (nturinski) [reply azure/index.d.ts:365]: I think the idea for it to be undefined was so that you could create a generic client to make a one-off request to an endpoint without having to specify a client sdk info.

Comment (alexweininger) [reply azure/index.d.ts:365]: Yeah but isn't my suggestion just changing the syntax and not the result?

Comment (nturinski) [reply azure/index.d.ts:365]: Oh yeah, I didn't see the ?.

Which leads me to why I'd prefer not to do your suggestion. When you mark it as optional, it makes omitting `clientInfo` less intentional. 

Whereas with the old syntax, I believe that you _must_ type `undefined` in the parameter list. It kind of enforces whoever is using the command to acknowledge that they are not passing in anything on purpose.

Comment (alexweininger) [reply azure/index.d.ts:365]: Nice, I guess that's a perfect scenario to not use `?`. 👍

File: azure/src/createAzureClient.ts [hunk 16]
```diff
@@ -208,3 +210,29 @@ class AddEndpointPolicy implements PipelinePolicy {
         return await next(request);
     }
 }
+
+/**
+ * The Azure SDK will only throw errors for bad status codes if it has an "operationSpec", but none of our "generic" requests will have that
+ */
+class StatusCodePolicy implements PipelinePolicy {
```
Comment (nturinski) [inline azure/src/createAzureClient.ts:217]: This is basically what we had before for the following reason:
> The Azure SDK will only throw errors for bad status codes if it has an "operationSpec", but none of our "generic" requests will have that

OperationSpec isn't something that you can pass into the request anymore, but I believe it's built internally into the SDK. If the SDK call errors out, it throws the error during the pipeline. However, with generic requests, it will accept any status code as a resolved response, so we need to be checking the error and throwing it ourselves.

File: azure/test/request.test.ts [hunk 17]
```diff
@@ -52,85 +52,66 @@ suite('request', () => {
...
     test('400 with json error message', async () => {
-        await assertThrowsAsync(async () => await sendTestRequest({ status: 400, body: '{ "message": "oops" }' }), (err: Error) => err.message.includes('oops') && !err.message.includes('message'));
+        await assertThrowsAsync(async () => await sendTestRequest({ statusCode: 400, body: '{ "message": "oops" }' }), (err: Error) => err.message.includes('oops') && !err.message.includes('message'));
     });
 
     test('ECONNRESET', async () => {
         await assertThrowsAsync(async () => await sendTestRequest(res => res.destroy()), /socket hang up/i);
     });
-
-    // test('operationSpec with unexpected  error', async () => {
```
Comment (nturinski) [inline azure/test/request.test.ts]: As I mentioned, we don't have `operationSpec` as a property anymore, so I removed these tests. The errors above basically test the same thing, which was the `StatusCodePolicy`

**Review by nturinski — COMMENTED** [review]
Okay, I think I am done with all the PR feedback and other things that I wanted to test. So shouldn't be adding anymore changes to this one (unless there ends up being more feedback)

**Comment by alexweininger** [comment]
Merge conflicts?

**Comment by nturinski** [comment]
Okaaay, _that_ is my last commit. Stupid version changes-- why can't git figure it out?

**Review by alexweininger — DISMISSED** [review]

**Review by alexweininger — APPROVED** [review]

**Comment by nturinski** [comment]
Need to release dev -> utils -> azure in that order to make fix the linting errors
